File: cb/db/repositories/profile_repo.py
"""Profile and token repository — SQLite3 CRUD."""

from __future__ import annotations
import time
from pathlib import Path

from cb.db.connection import get_connection
from cb.dtos.auth import ProfileDTO


# ── Profile repo ──────────────────────────────────────────────


def save_profile(
    name: str,
    server_url: str,
    username: str,
    is_default: bool = False,
    db_path: Path | None = None,
) -> ProfileDTO:
    now = int(time.time())
    conn = get_connection(db_path)
    try:
        if is_default:
            conn.execute("UPDATE profiles SET is_default = 0")

        # UPDATE existing profile (preserves id so token FK stays valid)
        cur = conn.execute(
            """UPDATE profiles
               SET server_url = ?, username = ?, is_default = ?
               WHERE name = ?""",
            (server_url.rstrip("/"), username, int(is_default), name),
        )
        if cur.rowcount == 0:
            # New profile — INSERT
            conn.execute(
                """INSERT INTO profiles
                   (name, server_url, username, is_default, created_at)
                   VALUES (?, ?, ?, ?, ?)""",
                (name, server_url.rstrip("/"), username, int(is_default), now),
            )
        conn.commit()
        return get_profile(name, db_path)
    finally:
        conn.close()


def get_profile(name: str, db_path: Path | None = None) -> ProfileDTO:
    conn = get_connection(db_path)
    try:
        row = conn.execute(
            "SELECT * FROM profiles WHERE name = ?", (name,)
        ).fetchone()
        if row is None:
            raise ValueError(f"Profile '{name}' not found.")
        return ProfileDTO(
            id=row["id"],
            name=row["name"],
            server_url=row["server_url"],
            username=row["username"],
            is_default=bool(row["is_default"]),
            created_at=row["created_at"],
        )
    finally:
        conn.close()


def get_default_profile(db_path: Path | None = None) -> ProfileDTO | None:
    conn = get_connection(db_path)
    try:
        row = conn.execute(
            "SELECT * FROM profiles WHERE is_default = 1 LIMIT 1"
        ).fetchone()
        if row is None:
            # Fall back to the first profile created
            row = conn.execute(
                "SELECT * FROM profiles ORDER BY created_at LIMIT 1"
            ).fetchone()
        if row is None:
            return None
        return ProfileDTO(
            id=row["id"],
            name=row["name"],
            server_url=row["server_url"],
            username=row["username"],
            is_default=bool(row["is_default"]),
            created_at=row["created_at"],
        )
    finally:
        conn.close()


def list_profiles(db_path: Path | None = None) -> list[ProfileDTO]:
    conn = get_connection(db_path)
    try:
        rows = conn.execute(
            "SELECT * FROM profiles ORDER BY is_default DESC, created_at"
        ).fetchall()
        return [
            ProfileDTO(
                id=r["id"],
                name=r["name"],
                server_url=r["server_url"],
                username=r["username"],
                is_default=bool(r["is_default"]),
                created_at=r["created_at"],
            )
            for r in rows
        ]
    finally:
        conn.close()


def delete_profile(name: str, db_path: Path | None = None) -> None:
    conn = get_connection(db_path)
    try:
        conn.execute("DELETE FROM profiles WHERE name = ?", (name,))
        conn.commit()
    finally:
        conn.close()

File: src/core/db/repositories/profile-repo.ts
/**
 * Profile repository — SQLite3 CRUD.
 * Behavioural 1:1 port of legacy/cb/db/repositories/profile_repo.py
 */

import { getConnection } from "../connection";
import { NotFoundError } from "../../api/errors";

export interface Profile {
  id: number;
  name: string;
  serverUrl: string;
  username: string;
  isDefault: boolean;
  createdAt: number;
}

// Raw row shape returned by bun:sqlite
interface ProfileRow {
  id: number;
  name: string;
  server_url: string;
  username: string;
  is_default: number;
  created_at: number;
}

function rowToProfile(row: ProfileRow): Profile {
  return {
    id: row.id,
    name: row.name,
    serverUrl: row.server_url,
    username: row.username,
    isDefault: row.is_default !== 0,
    createdAt: row.created_at,
  };
}

/**
 * Insert or update a profile. If is_default is true, clears is_default on all
 * other profiles first. Preserves the existing row id on update so any FK
 * relationships stay valid. Returns the saved profile.
 * Mirrors Python save_profile().
 */
export function saveProfile(
  name: string,
  serverUrl: string,
  username: string,
  isDefault = false,
  dbPath?: string
): Profile {
  const now = Math.floor(Date.now() / 1000);
  const db = getConnection(dbPath);
  try {
    if (isDefault) {
      db.run("UPDATE profiles SET is_default = 0");
    }

    // Try UPDATE first — preserves id so token FK stays valid
    const updated = db.run(
      `UPDATE profiles SET server_url = ?, username = ?, is_default = ? WHERE name = ?`,
      [serverUrl.replace(/\/+$/, ""), username, isDefault ? 1 : 0, name]
    );

    if (updated.changes === 0) {
      // New profile — INSERT
      db.run(
        `INSERT INTO profiles (name, server_url, username, is_default, created_at)
         VALUES (?, ?, ?, ?, ?)`,
        [name, serverUrl.replace(/\/+$/, ""), username, isDefault ? 1 : 0, now]
      );
    }
  } finally {
    db.close();
  }

  // getProfile opens its own connection — matches Python behaviour
  return getProfile(name, dbPath);
}

/**
 * Fetch a single profile by name. Throws if not found.
 * Mirrors Python get_profile().
 */
export function getProfile(name: string, dbPath?: string): Profile {
  const db = getConnection(dbPath);
  try {
    const row = db.query<ProfileRow, [string]>(
      "SELECT * FROM profiles WHERE name = ?"
    ).get(name);
    if (row === null) {
      throw new NotFoundError(`Profile '${name}' not found.`);
    }
    return rowToProfile(row);
  } finally {
    db.close();
  }
}

/**
 * Return the default profile, falling back to the oldest profile if none is
 * marked default. Returns null if no profiles exist.
 * Mirrors Python get_default_profile().
 */
export function getDefaultProfile(dbPath?: string): Profile | null {
  const db = getConnection(dbPath);
  try {
    let row = db.query<ProfileRow, []>(
      "SELECT * FROM profiles WHERE is_default = 1 LIMIT 1"
    ).get();

    if (row === null) {
      // Fall back to the first profile created
      row = db.query<ProfileRow, []>(
        "SELECT * FROM profiles ORDER BY created_at LIMIT 1"
      ).get();
    }

    return row !== null ? rowToProfile(row) : null;
  } finally {
    db.close();
  }
}

/**
 * Return all profiles ordered by default-first then creation time.
 * Mirrors Python list_profiles().
 */
export function listProfiles(dbPath?: string): Profile[] {
  const db = getConnection(dbPath);
  try {
    const rows = db.query<ProfileRow, []>(
      "SELECT * FROM profiles ORDER BY is_default DESC, created_at"
    ).all();
    return rows.map(rowToProfile);
  } finally {
    db.close();
  }
}

/**
 * Delete a profile by name.
 * Mirrors Python delete_profile().
 */
export function deleteProfile(name: string, dbPath?: string): void {
  const db = getConnection(dbPath);
  try {
    db.run("DELETE FROM profiles WHERE name = ?", [name]);
  } finally {
    db.close();
  }
}
